sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java

```diff
@@ -594,6 +607,182 @@ private void setRequestHeaders(CosmosChangeFeedRequestOptions requestOptions) {
+                            itemBatchOperation.getRequestOptions(),
+                            null
+                        )
+                    );
+            }
+            monoList.add(itemBatchOperationMono);
+        }
+        Mono<List<ItemBatchOperation<?>>> encryptedOperationListMono =
+            Flux.mergeSequential(monoList).collectList();
+        CosmosBatchRequestOptions finalRequestOptions = requestOptions;
+
+        CosmosBatch encryptedCosmosBatch = CosmosBatch.createCosmosBatch(cosmosBatch.getPartitionKeyValue());
+
+        return encryptedOperationListMono.flatMap(itemBatchOperations -> {
+            this.cosmosBatchAccessor.getOperationsInternal(encryptedCosmosBatch).addAll(itemBatchOperations);
+            return executeCosmosBatchHelper(encryptedCosmosBatch, finalRequestOptions, false);
+        });
+    }
+
+    private Mono<CosmosBatchResponse> executeCosmosBatchHelper(CosmosBatch encryptedCosmosBatch,
+                                                               CosmosBatchRequestOptions requestOptions,
+                                                               boolean isRetry) {
+        setRequestHeaders(requestOptions);
+        return this.container.executeCosmosBatch(encryptedCosmosBatch, requestOptions).flatMap(cosmosBatchResponse -> {
+            // TODO this should check for BadRequest StatusCode too, requires a service fix to return 400 instead of
+            //  -1 which is currently returned inside the body.
+            //  Once fixed from service below if condition can be removed, as this is already covered in onErrorResume.
+            if (!isRetry && cosmosBatchResponse.getSubStatusCode() == Integer.valueOf(Constants.INCORRECT_CONTAINER_RID_SUB_STATUS)) {
+                this.encryptionProcessor.getIsEncryptionSettingsInitDone().set(false);
+                return this.encryptionProcessor.initializeEncryptionSettingsAsync(true).then
+                    (Mono.defer(() -> executeCosmosBatchHelper(encryptedCosmosBatch, requestOptions, true)));
+            }
+
+            List<Mono<Void>> decryptMonoList = new ArrayList<>();
+            for (CosmosBatchOperationResult cosmosBatchOperationResult :
+                this.cosmosBatchResponseAccessor.getResults(cosmosBatchResponse)) {
+                ObjectNode objectNode =
+                    this.cosmosBatchOperationResultAccessor.getResourceObject(cosmosBatchOperationResult);
+                if (objectNode != null) {
+                    decryptMonoList.add(encryptionProcessor.decryptObjectNode(objectNode).flatMap(jsonNode -> {
+                        this.cosmosBatchOperationResultAccessor.setResourceObject(cosmosBatchOperationResult, jsonNode);
+                        return Mono.empty();
+                    }));
+                }
+            }
+
+            Mono<List<Void>> listMono = Flux.mergeSequential(decryptMonoList).collectList();
+            return listMono.flatMap(aVoid -> Mono.just(cosmosBatchResponse));
```

Comment:
is this guaranteed that `listMono` is never `Mono.empty`?

Reply:
Yes listMono is never Mono.empty. Tested it by single batch with delete operation , also by  manually passing  List<Mono<Void>> decryptMonoList = new ArrayList<>() directly